src/main/java/net/minestom/server/network/player/PlayerConnection.java

```diff
@@ -119,8 +121,12 @@ public void disconnect() {
         this.online = false;
         MinecraftServer.getConnectionManager().removePlayer(this);
         final Player player = getPlayer();
-        if (player != null && !player.isRemoved()) {
-            player.scheduleNextTick(Entity::remove);
+        if (player != null) {
+            if (connectionState == ConnectionState.PLAY && !player.isRemoved())
+                player.scheduleNextTick(Entity::remove);
+            else {
+                EventDispatcher.call(new PlayerDisconnectEvent(player));
```

Comment:
This should only be called if player is not removed, and should call remove itself. Currently if 3 different handlers of `AsyncPlayerConfigurationEvent` kicked the player, `PlayerDisconnectEvent` would be called 3 times.